File: backend/Nutrition-backend/src/models/schemas/supplement_comprehensive.py
"""5-card 종합 분석 (B-persona 차별화) Pydantic schema.

`POST /api/v1/supplements/analyze/comprehensive` endpoint 의 입출력. OCR analyze
endpoint 가 반환하는 ingredient candidate 들을 KDRIs + 만성질환 매트릭스와
교차하여 모바일 5-card UI 의 5종 카드를 모두 채울 수 있도록 한다:

- 카드 1: 부족 영양소 (`deficient_nutrients`)
- 카드 2: 과다 섭취 (`excessive_nutrients`)
- 카드 3: 주의 성분 (`cautionary_components`)
- 카드 4: 식단/영양제 점수 (`diet_score`)
- 카드 5: 목적별 / 만성질환 (`purpose_targets`)

Reference:
    outputs/todo-list/2026-05-21/b-persona-accuracy-report.md §3
    data/nutrition_reference/chronic_disease_supplement_matrix.json
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from src.models.schemas.chronic_disease_matrix import ChronicCondition

PersonaTag = Literal["A", "B"]
"""사용자 페르소나. A=박직장(예방), B=김건강(만성질환자, 차별화 핵심)."""

ScoreLabel = Literal["excellent", "good", "moderate", "warning", "critical"]
"""식단 점수의 5단계 자연어 라벨."""

Severity = Literal["low", "medium", "high"]
"""주의 성분의 심각도 등급."""


class ComprehensiveIngredient(BaseModel):
    """클라이언트가 보낸 ingredient 정보 (analyze endpoint 응답에서 추출).

    Attributes:
        display_name: 사용자에게 표시되는 이름 (한글/영문 그대로).
        nutrient_code: 내부 영양소 코드 (예: `vitamin_d_ug`). 매트릭스 룩업용.
        amount: 1회 섭취량.
        unit: 단위 (예: `mg`, `ug`, `IU`).
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    display_name: str = Field(min_length=1, max_length=120)
    nutrient_code: str | None = Field(default=None, max_length=80)
    amount: float | None = Field(default=None, ge=0)
    unit: str | None = Field(default=None, max_length=40)


class UserProfileInput(BaseModel):
    """KDRIs 룩업과 만성질환 매핑에 필요한 최소 사용자 프로필.

    Attributes:
        age: 만 나이 (1~120).
        sex: 성별.
        chronic_conditions: 사용자가 보유한 만성질환 인디케이션.
        is_pregnant: 임신 여부 (KDRIs 분기에 사용, 선택).
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    age: int = Field(ge=1, le=120)
    sex: Literal["male", "female"]
    chronic_conditions: list[ChronicCondition] = Field(default_factory=list, max_length=8)
    is_pregnant: bool = Field(default=False)


class ComprehensiveAnalysisRequest(BaseModel):
    """`POST /api/v1/supplements/analyze/comprehensive` 요청 본문.

    Attributes:
        analysis_id: 선행 OCR analyze 의 식별자 (감사 로그 용도).
        ingredients: 분석 대상 영양제 ingredient 리스트.
        user_profile: KDRIs 룩업용 사용자 프로필.
        persona: A/B 페르소나 분기 (점수 가중치 조정용).
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    analysis_id: str | None = Field(default=None, max_length=80)
    ingredients: list[ComprehensiveIngredient] = Field(min_length=0, max_length=80)
    user_profile: UserProfileInput
    persona: PersonaTag = "B"


class DeficientNutrient(BaseModel):
    """부족 영양소 (KDRIs 권장량 미달).

    Attributes:
        nutrient_code: 내부 영양소 코드.
        display_name: 사용자 표시명.
        current_intake: 현재 섭취량.
        recommended_intake: KDRIs 일일 권장량.
        unit: 비교 단위.
        deficit_ratio: 부족 비율 0.0~1.0 (1.0 = 100% 부족).
    """

    model_config = ConfigDict(extra="forbid")

    nutrient_code: str
    display_name: str
    current_intake: float
    recommended_intake: float
    unit: str
    deficit_ratio: float = Field(ge=0.0, le=1.0)


class ExcessiveNutrient(BaseModel):
    """과다 섭취 영양소 (KDRIs UL 초과).

    Attributes:
        nutrient_code: 내부 영양소 코드.
        display_name: 사용자 표시명.
        current_intake: 현재 섭취량.
        upper_limit: KDRIs 상한 섭취량 (UL).
        unit: 비교 단위.
        excess_ratio: 초과 비율 (1.5 = UL 의 1.5 배).
    """

    model_config = ConfigDict(extra="forbid")

    nutrient_code: str
    display_name: str
    current_intake: float
    upper_limit: float
    unit: str
    excess_ratio: float = Field(ge=1.0)


class CautionaryComponent(BaseModel):
    """주의 성분 (약물 상호작용, 만성질환 회피군 등).

    Attributes:
        component: 성분/카테고리 이름.
        reason: `drug_interaction:warfarin` 등 표준 토큰.
        severity: 심각도.
        message: 사용자 친화 한국어 안내.
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    component: str
    reason: str
    severity: Severity
    message: str


class PurposeTarget(BaseModel):
    """만성질환 인디케이션별 적합도 (B-persona 차별화 핵심).

    Attributes:
        condition: 만성질환 인디케이션.
        relevance_score: 0.0~1.0 본 supplement 의 적합도.
        evidence_level: EBM 등급.
        message: 사용자 친화 한국어 안내.
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    condition: ChronicCondition
    relevance_score: float = Field(ge=0.0, le=1.0)
    evidence_level: Literal["strong", "moderate", "weak", "insufficient"]
    message: str


class SupplementComprehensiveAnalysis(BaseModel):
    """`POST /api/v1/supplements/analyze/comprehensive` 응답.

    Attributes:
        analysis_id: 요청에서 전달된 식별자 echo.
        persona: 적용된 페르소나.
        deficient_nutrients: 카드 1.
        excessive_nutrients: 카드 2.
        cautionary_components: 카드 3.
        diet_score: 0~100 종합 점수 (카드 4).
        diet_score_label: 점수 라벨.
        diet_score_message: 사용자 친화 한국어 한 줄 코멘트.
        purpose_targets: 만성질환 인디케이션별 적합도 (카드 5).
        chronic_disease_indications: matrix 기반 자동 매핑 결과.
        algorithm_version: 산출 로직 버전 (회귀 추적용).
        warnings: 산출 과정에서 발생한 경고.
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    analysis_id: str | None = None
    persona: PersonaTag
    deficient_nutrients: list[DeficientNutrient] = Field(default_factory=list)
    excessive_nutrients: list[ExcessiveNutrient] = Field(default_factory=list)
    cautionary_components: list[CautionaryComponent] = Field(default_factory=list)
    diet_score: int = Field(ge=0, le=100)
    diet_score_label: ScoreLabel
    diet_score_message: str
    purpose_targets: list[PurposeTarget] = Field(default_factory=list)
    chronic_disease_indications: list[ChronicCondition] = Field(default_factory=list)
    algorithm_version: str = "comprehensive-v1"
    warnings: list[str] = Field(default_factory=list)

File: mobile/lib/models/supplement_comprehensive.dart
/// Backend `POST /api/v1/supplements/analyze/comprehensive` 의 입출력을 위한
/// Dart 모델. SupplementAnalysisPreview 의 후속 호출로 5-card UI 의 5종 카드
/// 데이터를 모두 채운다.
library;

import 'package:flutter/foundation.dart';

/// 만성질환 인디케이션 토큰 (Pydantic ChronicCondition Literal 과 일치).
typedef ChronicCondition = String;

/// `ComprehensiveIngredient` 입력 구조.
@immutable
class ComprehensiveIngredientPayload {
  const ComprehensiveIngredientPayload({
    required this.displayName,
    this.nutrientCode,
    this.amount,
    this.unit,
  });

  final String displayName;
  final String? nutrientCode;
  final double? amount;
  final String? unit;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'display_name': displayName,
        if (nutrientCode != null) 'nutrient_code': nutrientCode,
        if (amount != null) 'amount': amount,
        if (unit != null) 'unit': unit,
      };
}

/// `UserProfileInput` 입력 구조.
@immutable
class UserProfilePayload {
  const UserProfilePayload({
    required this.age,
    required this.sex,
    this.chronicConditions = const <ChronicCondition>[],
    this.isPregnant = false,
  });

  final int age;
  final String sex; // 'male' | 'female'
  final List<ChronicCondition> chronicConditions;
  final bool isPregnant;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'age': age,
        'sex': sex,
        'chronic_conditions': chronicConditions,
        'is_pregnant': isPregnant,
      };
}

@immutable
class DeficientNutrient {
  const DeficientNutrient({
    required this.nutrientCode,
    required this.displayName,
    required this.currentIntake,
    required this.recommendedIntake,
    required this.unit,
    required this.deficitRatio,
  });

  final String nutrientCode;
  final String displayName;
  final double currentIntake;
  final double recommendedIntake;
  final String unit;
  final double deficitRatio;

  factory DeficientNutrient.fromJson(Map<String, dynamic> json) {
    return DeficientNutrient(
      nutrientCode: json['nutrient_code'] as String? ?? '',
      displayName: json['display_name'] as String? ?? '',
      currentIntake: (json['current_intake'] as num?)?.toDouble() ?? 0.0,
      recommendedIntake: (json['recommended_intake'] as num?)?.toDouble() ?? 0.0,
      unit: json['unit'] as String? ?? '',
      deficitRatio: (json['deficit_ratio'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

@immutable
class ExcessiveNutrient {
  const ExcessiveNutrient({
    required this.nutrientCode,
    required this.displayName,
    required this.currentIntake,
    required this.upperLimit,
    required this.unit,
    required this.excessRatio,
  });

  final String nutrientCode;
  final String displayName;
  final double currentIntake;
  final double upperLimit;
  final String unit;
  final double excessRatio;

  factory ExcessiveNutrient.fromJson(Map<String, dynamic> json) {
    return ExcessiveNutrient(
      nutrientCode: json['nutrient_code'] as String? ?? '',
      displayName: json['display_name'] as String? ?? '',
      currentIntake: (json['current_intake'] as num?)?.toDouble() ?? 0.0,
      upperLimit: (json['upper_limit'] as num?)?.toDouble() ?? 0.0,
      unit: json['unit'] as String? ?? '',
      excessRatio: (json['excess_ratio'] as num?)?.toDouble() ?? 1.0,
    );
  }
}

@immutable
class CautionaryComponent {
  const CautionaryComponent({
    required this.component,
    required this.reason,
    required this.severity,
    required this.message,
  });

  final String component;
  final String reason;
  final String severity; // 'low' | 'medium' | 'high'
  final String message;

  factory CautionaryComponent.fromJson(Map<String, dynamic> json) {
    return CautionaryComponent(
      component: json['component'] as String? ?? '',
      reason: json['reason'] as String? ?? '',
      severity: json['severity'] as String? ?? 'low',
      message: json['message'] as String? ?? '',
    );
  }
}

@immutable
class PurposeTarget {
  const PurposeTarget({
    required this.condition,
    required this.relevanceScore,
    required this.evidenceLevel,
    required this.message,
  });

  final ChronicCondition condition;
  final double relevanceScore;
  final String evidenceLevel; // 'strong' | 'moderate' | 'weak' | 'insufficient'
  final String message;

  factory PurposeTarget.fromJson(Map<String, dynamic> json) {
    return PurposeTarget(
      condition: json['condition'] as String? ?? '',
      relevanceScore: (json['relevance_score'] as num?)?.toDouble() ?? 0.0,
      evidenceLevel: json['evidence_level'] as String? ?? 'insufficient',
      message: json['message'] as String? ?? '',
    );
  }
}

@immutable
class SupplementComprehensiveAnalysis {
  const SupplementComprehensiveAnalysis({
    required this.persona,
    required this.deficientNutrients,
    required this.excessiveNutrients,
    required this.cautionaryComponents,
    required this.dietScore,
    required this.dietScoreLabel,
    required this.dietScoreMessage,
    required this.purposeTargets,
    required this.chronicDiseaseIndications,
    required this.algorithmVersion,
    required this.warnings,
    this.analysisId,
  });

  final String? analysisId;
  final String persona; // 'A' | 'B'
  final List<DeficientNutrient> deficientNutrients;
  final List<ExcessiveNutrient> excessiveNutrients;
  final List<CautionaryComponent> cautionaryComponents;
  final int dietScore; // 0~100
  final String dietScoreLabel; // 'excellent' | 'good' | 'moderate' | 'warning' | 'critical'
  final String dietScoreMessage;
  final List<PurposeTarget> purposeTargets;
  final List<ChronicCondition> chronicDiseaseIndications;
  final String algorithmVersion;
  final List<String> warnings;

  factory SupplementComprehensiveAnalysis.fromJson(Map<String, dynamic> json) {
    return SupplementComprehensiveAnalysis(
      analysisId: json['analysis_id'] as String?,
      persona: json['persona'] as String? ?? 'B',
      deficientNutrients: ((json['deficient_nutrients'] as List?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(DeficientNutrient.fromJson)
          .toList(growable: false),
      excessiveNutrients: ((json['excessive_nutrients'] as List?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(ExcessiveNutrient.fromJson)
          .toList(growable: false),
      cautionaryComponents: ((json['cautionary_components'] as List?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(CautionaryComponent.fromJson)
          .toList(growable: false),
      dietScore: (json['diet_score'] as num?)?.toInt() ?? 0,
      dietScoreLabel: json['diet_score_label'] as String? ?? 'moderate',
      dietScoreMessage: json['diet_score_message'] as String? ?? '',
      purposeTargets: ((json['purpose_targets'] as List?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map(PurposeTarget.fromJson)
          .toList(growable: false),
      chronicDiseaseIndications: ((json['chronic_disease_indications'] as List?) ?? const [])
          .whereType<String>()
          .toList(growable: false),
      algorithmVersion: json['algorithm_version'] as String? ?? 'unknown',
      warnings: ((json['warnings'] as List?) ?? const [])
          .whereType<String>()
          .toList(growable: false),
    );
  }
}
